### PR TITLE
[GSOC] feat: Added Integration Testing

### PR DESCRIPTION
**Pull request by syedali237**
## Description

This PR introduces integration testing for the application as part of the ongoing GSoC development process. The goal is to improve the reliability and stability of the codebase by ensuring that various components interact correctly under real-world scenarios.

## Changes Made

- Implemented integration tests to validate end-to-end workflows.
- Enhanced error handling in `chat_service` for better exception handlings and debugging.
- Fixed the **Floating Action Button (FAB)** navigation issue by adding unique `Hero` tags to avoid animation conflicts during screen transitions.

A video demonstration has been added in a further comment.

## Summary by CodeRabbit

* **New Features**
  * Introduced end-to-end integration tests for admin login and logout flows.
  * Added helper utilities and robot classes to automate login and logout actions in tests.

* **Bug Fixes**
  * Improved error handling and fallback logic for fetching attachment URLs, ensuring a valid placeholder image is used when needed.
  * Enhanced chat service error handling to gracefully handle missing authentication and user IDs.

* **Improvements**
  * Added explicit hero tags to floating action buttons across multiple screens to support better navigation animations.
  * Updated main screen to use a fixed scaffold key for consistency.
  * Enabled test environment configuration for GraphQL services.

* **Chores**
  * Added the integration_test package as a development dependency.
  * Added an empty helper file for HTTP overrides in integration tests.

### CODE REVIEW

**Comment by syedali237** [comment]
@may-tas for your reference

File: lib/services/chat_service.dart [hunk 1]
```diff
@@ -140,29 +144,52 @@ class ChatService {
...
+  if (_userConfig.currentUser.authToken == null || 
+      _userConfig.currentUser.authToken!.isEmpty) {
+    debugPrint('Skipping chat fetch: Authentication token not available yet');
+    return [];
+  }
+
     final userId = _userConfig.currentUser.id;
 
     if (userId == null) {
-      navigationService.showTalawaErrorDialog(
```
Comment (may-tas) [inline lib/services/chat_service.dart:148]: hey @syedali237 why was this dialog and throw statement removed.

Comment (syedali237) [reply lib/services/chat_service.dart:148]: this just pop-ups a dialog error that can be dismissed by the user and further chat functionality is working properly

Comment (may-tas) [reply lib/services/chat_service.dart:148]: Yes so its fine to keep it, why did we had to remove it?

Comment (may-tas) [reply lib/services/chat_service.dart:148]: if the userId == null then the debugPrint will work for the developers, but the user won't know what went wrong, can you please add it back.

Comment (syedali237) [reply lib/services/chat_service.dart:148]: just refactored it, it still shows a dialog box with that error when you open the application and it shows it to the user to dismiss it

Comment (may-tas) [reply lib/services/chat_service.dart:148]: But there is no dialog being shown for userId==null.
Pls check again

Comment (syedali237) [reply lib/services/chat_service.dart:148]: ok i'll check

File: integration_test/robots/login_form_robot.dart [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    final Finder dismissButton = find.text('Dismiss');
+    if (dismissButton.evaluate().isNotEmpty) {
+      print("Dismiss button found, tapping it");
+      await tester.tap(dismissButton);
+      await tester.pumpAndSettle(const Duration(seconds: 2));
+    } else {
+      print("Dismiss button not found, continuing...");
+    }
+    final scaffoldFinder = find.byKey(const Key('MainScaffold'));
+    // final scaffoldFinder = find.byType(Scaffold);
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:82]: done in the future commit

**Comment by syedali237** [comment]
https://github.com/user-attachments/assets/ea249909-2c2a-44ed-b59d-058eb893434f


https://github.com/user-attachments/assets/f241749b-730a-4b28-becb-af3a2c3d4f7c

File: lib/services/chat_service.dart (hunk 1 above)
Comment (syedali237) [reply lib/services/chat_service.dart:148]: @may-tas I have added back the error dialog

File: lib/services/chat_service.dart [hunk 3]
```diff
@@ -140,29 +144,53 @@
...
     if (userId == null) {
+      debugPrint('Error fetching chats: User ID is required');
       navigationService.showTalawaErrorDialog(
-        "User ID is required to fetch chats",
-        MessageType.error,
-      );
-      throw ArgumentError('User ID is required to fetch chats');
+          "User ID is requires to fetch chats", MessageType.error,);
+      return [];
     }
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/services/chat_service.dart]: resolved in further commit

File: lib/services/graphql_config.dart [hunk 4]
```diff
@@ -29,6 +29,12 @@
     getOrgUrl();
     return true;
   }
+  
+  /// This function is used to initialize the GraphQL client for testing.
+  void initializeForTesting(String apiUrl) {
+    httpLink = HttpLink(apiUrl);
+    orgURI = apiUrl;
+  }
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/services/graphql_config.dart:37]: added in furtther commit

File: test/model_tests/post/post_model_test.dart [hunk 5]
```diff
@@ -222,5 +225,107 @@
...
+        ),
+      );
+
+      await post.getPresignedUrl('org1');
+
+      expect(
+        post.attachments!.first.url,
+        'https://avatars.githubusercontent.com/u/24500036?s=280&v=4',
+      );
+    });
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/model_tests/post/post_model_test.dart]: for now this approach is fine

File: test/service_tests/chat_service_test.dart [hunk 6]
```diff
@@ -42,6 +42,7 @@ void main() {
         firstName: 'John',
         lastName: 'Doe',
         email: 'john@example.com',
+        authToken: 'sample-auth-token',
       );
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/service_tests/chat_service_test.dart:47]: added in further commit

File: test/service_tests/chat_service_test.dart [hunk 7]
```diff
@@ -311,44 +313,102 @@ void main() {
...
-          throwsA(
-            isA<Exception>().having(
-              (e) => e.toString(),
-              'toString',
-              contains('Failed to fetch chats'),
+        final result = await chatService.getChatsByUser();
+
+        expect(result, isNotNull);
+        expect(result, isEmpty);
+      });
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/service_tests/chat_service_test.dart:368]: This test is intentionally verifying that authentication failures return an empty list.

File: test/service_tests/graphql_config_test.dart [hunk 8]
```diff
@@ -204,6 +204,17 @@ void main() {
...
+      const testApiUrl = 'https://test-api.example.com/graphql';
+      final config = GraphqlConfig();
+
+      config.initializeForTesting(testApiUrl);
+
+      expect(config.httpLink, isA<HttpLink>());
+      expect(config.httpLink.uri.toString(), testApiUrl);
+      expect(GraphqlConfig.orgURI, testApiUrl);
+    });
+
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/service_tests/graphql_config_test.dart:217]: This isn't necessary.

File: test/model_tests/post/post_model_test.dart [hunk 9]
```diff
@@ -222,5 +225,107 @@
...
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/model_tests/post/post_model_test.dart]: for now the required tests already covers the edge cases

File: integration_test/app_test.dart [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        devPrint("✓ Admin logged in successfully");
+        logoutRobot = LogoutRobot(tester);
+        devPrint("-> Logout Admin");
+        await logoutRobot.logout();
+        devPrint("✓ Admin logged out successfully");
+
+        devPrint("✓ Admin test completed successfully");
+      },
+      timeout: const Timeout(Duration(minutes: 5)),
+    );
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/app_test.dart]: resolved in future commit

File: integration_test/robots/login_form_robot.dart [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  ///
+  /// **params**:
+  ///   None
+  ///
+  /// **returns**:
+  ///   None
+  Future<void> loginAdmin() async {
+    await _selectLanguage();
+    await _openLoginForm();
+    await _enterEmailAndPassword("testadmin1@example.com", "Pass@123");
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:29]: this is done keeping in mind for integration test

File: integration_test/robots/login_form_robot.dart [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    await tester.pumpAndSettle(const Duration(seconds: 5));
+    final Finder selectLanguageButton =
+        find.byKey(const Key('SelectLangTextButton'));
+    await tester.pumpAndSettle();
+    await tester.pump(const Duration(milliseconds: 500));
+    expect(selectLanguageButton, findsOneWidget);
+    await tester.tap(selectLanguageButton);
+    await tester.pump(const Duration(milliseconds: 500));
+    await tester.tap(find.text('Skip'));
+    print("Language selected");
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart]: resolved in future commit

File: lib/services/chat_service.dart [hunk 13]
```diff
@@ -140,29 +144,51 @@ class ChatService {
...
+
+      // For other errors, still return empty list but could optionally show error dialog
+      // Only show error dialog if we're not in a test environment
+      if (!exceptionString.contains('test')) {
+        navigationService.showTalawaErrorDialog(
+          "Failed to fetch chats",
+          MessageType.error,
+        );
+      }
+      return [];
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/services/chat_service.dart:190]: this meets our current requirements

File: lib/services/chat_service.dart [hunk 14]
```diff
@@ -140,29 +144,53 @@ class ChatService {
   ///   None
   ///
   /// **returns**:
-  /// * `Future<List<Chat>?>`: True if successful, false otherwise
+  /// * `Future<List<Chat>?>`: List of chats if successful, empty list if authentication fails, null if other errors
   Future<List<Chat>?> getChatsByUser() async {
+    if (_userConfig.currentUser.authToken == null ||
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/services/chat_service.dart]: done in future commit

File: test/service_tests/chat_service_test.dart [hunk 15]
```diff
@@ -311,44 +313,102 @@
...
+              ],
             ),
           ),
         );
+
+        final result = await chatService.getChatsByUser();
+
+        expect(result, isNotNull);
+        expect(result, isEmpty);
+      });
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/service_tests/chat_service_test.dart:389]: the existing code, already ensures that no dialog appears

File: test/model_tests/post/post_model_test.dart [hunk 16]
```diff
@@ -202,7 +202,10 @@ void main() {
 
       await post.getPresignedUrl('org1');
 
-      expect(post.attachments!.first.url, isNull);
+      expect(
+        post.attachments!.first.url,
+        'https://avatars.githubusercontent.com/u/24500036?s=280&v=4',
+      );
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/model_tests/post/post_model_test.dart]: added in future commit

File: test/model_tests/post/post_model_test.dart [hunk 17]
```diff
@@ -222,5 +225,107 @@ void main() {
...
+          .thenThrow(Exception('Network error'));
+
+      await post.getPresignedUrl('org1');
+
+      // Should set fallback URL when exception occurs
+      expect(
+        post.attachments!.first.url,
+        'https://avatars.githubusercontent.com/u/24500036?s=280&v=4',
+      );
+    });
```
[the remark this answers is not in the document]
Comment (syedali237) [reply test/model_tests/post/post_model_test.dart]: for now our test covers our requirements

File: lib/views/main_screen.dart [hunk 18]
```diff
@@ -34,7 +34,7 @@ class _MainScreenState extends State<MainScreen> {
         // Setup navigation items and pages for the bottom navigation bar
         model.setupNavigationItems(context);
         return Scaffold(
-          key: model.scaffoldKey,
+          key: const Key('MainScaffold'),
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/views/main_screen.dart:37]: for now this works fine

File: integration_test/robots/login_form_robot.dart [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+    await _openLoginForm();
+    await _enterEmailAndPassword("testadmin1@example.com", "Pass@123");
+    await _submit();
+    final orgHeader = find.text("Unity Foundation - North");
+    await tester.pumpAndSettle(const Duration(seconds: 3));
+    expect(
+      orgHeader,
+      findsOneWidget,
+      reason: 'Admin login failed or page not loaded',
+    );
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:37]: its trying to get the actual title

File: lib/services/chat_service.dart [hunk 20]
```diff
@@ -38,6 +39,9 @@ class ChatService {
   /// User configuration instance.
   final _userConfig = locator<UserConfig>();
 
+  /// Navigation service instance.
+  final navigationService = locator<NavigationService>();
+
```
[the remark this answers is not in the document]
Comment (syedali237) [reply lib/services/chat_service.dart:44]: checks our current requirements

File: integration_test/app_test.dart [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+      "For Admin",
+      (WidgetTester tester) async {
+        app.main();
+
+        await tester.pumpAndSettle();
+
+        final graphqlConfig = locator<GraphqlConfig>();
+        graphqlConfig.initializeForTesting(dotenv.get('API_URL'));
+
+        await Future.delayed(const Duration(seconds: 2));
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/app_test.dart:37]: it has been deliberately done

File: integration_test/app_test.dart [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  setUpAll(() async {
+    await dotenv.load(fileName: ".env");
+
+    debugPrint("Environment variables loaded");
+
+    debugPrint("API_URL = ${dotenv.get('API_URL')}");
+  });
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/app_test.dart:24]: it loads fine with this setup

File: integration_test/robots/login_form_robot.dart [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+  Future<void> _selectLanguage() async {
+    await tester.pumpAndSettle(const Duration(seconds: 5));
+    final Finder selectLanguageButton =
+        find.byKey(const Key('SelectLangTextButton'));
+    await tester.pumpAndSettle();
+    await tester.pump(const Duration(milliseconds: 500));
+    expect(selectLanguageButton, findsOneWidget);
+    await tester.tap(selectLanguageButton);
+    await tester.pump(const Duration(milliseconds: 500));
+    await tester.tap(find.text('Skip'));
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:56]: resolved in future commit

File: integration_test/robots/login_form_robot.dart [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+  /// 2. Opens the login form.
+  /// 3. Enters the email and password for the admin user.
+  /// 4. Submits the login form.
+  ///
+  /// **params**:
+  ///   None
+  ///
+  /// **returns**:
+  ///   None
+  Future<void> loginAdmin() async {
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:26]: can be documented like this too

File: integration_test/robots/login_form_robot.dart [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+  /// **params**:
+  ///   None
+  ///
+  /// **returns**:
+  ///   None
+  Future<void> _openLoginForm() async {
+    await tester.pumpAndSettle();
+    final Finder dismissButton = find.text('Dismiss');
+    if (dismissButton.evaluate().isNotEmpty) {
+      devPrint("Dismiss button found, tapping it");
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:75]: current code works fine

File: integration_test/robots/login_form_robot.dart [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+    final ScaffoldState scaffoldState = tester.state(scaffoldFinder);
+    scaffoldState.openDrawer();
+
+    await tester.pumpAndSettle();
+    await tester.tap(find.widgetWithIcon(ListTile, Icons.add));
+    await tester.pumpAndSettle();
+    final Finder loginButton = find.text('Login');
+    await tester.tap(loginButton);
+    await tester.pumpAndSettle();
+  }
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:95]: works with all screens

File: integration_test/robots/login_form_robot.dart [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+    await tester.pumpAndSettle(const Duration(seconds: 5));
+    final Finder selectLanguageButton =
+        find.byKey(const Key('SelectLangTextButton'));
+    await tester.pumpAndSettle();
+    await tester.pump(const Duration(milliseconds: 500));
+    expect(selectLanguageButton, findsOneWidget);
+    await tester.tap(selectLanguageButton);
+    await tester.pump(const Duration(milliseconds: 500));
+    await tester.tap(find.text('Skip'));
+    devPrint("Language selected");
```
[the remark this answers is not in the document]
Comment (syedali237) [reply integration_test/robots/login_form_robot.dart:57]: done deliberately

**Review by Kevonia — APPROVED** [review]

**Review by noman2002 — APPROVED** [review]
LGTM
